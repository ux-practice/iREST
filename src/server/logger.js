import {format, createLogger, transports} from 'winston'

const {printf, combine, timestamp, label, colorize} = format

// eslint-disable-next-line no-shadow
export const customFormat = ({label, timestamp, level, message, ...rest}) => `${label} ${timestamp} ${level}: ${message} ${JSON.stringify(rest) === "{}" ? "" : JSON.stringify(rest)}`

const logger = createLogger(process.env.NODE_ENV === 'production' ? {transports: [ new transports.Console({level: 'error'})]}:{
  level: 'debug',
  transports: [
    new transports.Console({
      format: combine(colorize(), combine(
        colorize({
          all: true,
        }),
        label({
          label: '\n[Default Logger]',
        }),
        timestamp({
          format: 'DD-MM-YY HH:MM:SS',
        }),
        printf(customFormat)
      )),
    }),
  ],
})

export default logger
