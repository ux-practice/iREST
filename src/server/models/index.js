import User from './user'
import Project from './project'
import Mock from './mock'
import TempMock from './previewMock'
import ServiceResponse from './serviceResponse'
import TempServiceResponse from './previewServiceResponse'
import Token from './token'
import ApiResponseSchemaType from './ApiResponseSchemaTypes'
import TempApiResponseSchemaType from './previewApiResponseSchemaTypes'

const dbConn = (db, sequelize, DataTypes) => {
    db.User = User(sequelize, DataTypes)
    db.Mock = Mock(sequelize, DataTypes)
    db.TempMock = TempMock(sequelize, DataTypes)
    db.Project = Project(sequelize, DataTypes)
    db.ServiceResponse = ServiceResponse(sequelize, DataTypes)
    db.Token = Token(sequelize, DataTypes)
    db.TempServiceResponse = TempServiceResponse(sequelize, DataTypes)
    db.ApiResponseSchemaType = ApiResponseSchemaType(sequelize, DataTypes)
    db.TempApiResponseSchemaType = TempApiResponseSchemaType(sequelize, DataTypes)

    db.User.hasMany(db.Project, {
        foreignKey: 'userId',
        foreignKeyConstraint: true,
        onDelete: 'NO ACTION',
    })
    db.Project.belongsTo(db.User, {
        foreignKey: 'userId',
        foreignKeyConstraint: true,
        onDelete: 'NO ACTION',
    })
    db.User.hasMany(db.Mock, {
        foreignKey: 'userId',
        foreignKeyConstraint: true,
        onDelete: 'NO ACTION',
    })
    db.Project.hasMany(db.Mock, {
        foreignKey: 'projectId',
        foreignKeyConstraint: true,
        onDelete: 'NO ACTION',
    })
    db.Mock.belongsTo(db.User, {
        foreignKey: 'userId',
        foreignKeyConstraint: true,
        onDelete: 'NO ACTION',
    })
    db.Mock.belongsTo(db.Project, {
        foreignKey: 'projectId',
        foreignKeyConstraint: true,
        onDelete: 'NO ACTION',
    })
    db.Mock.hasMany(db.ServiceResponse, {
        foreignKey: 'mockId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.TempMock.belongsTo(db.User, {
        foreignKey: 'userId',
        foreignKeyConstraint: true,
        onDelete: 'NO ACTION',
    })
    db.TempMock.belongsTo(db.Project, {
        foreignKey: 'projectId',
        foreignKeyConstraint: true,
        onDelete: 'NO ACTION',
    })
    db.TempMock.hasMany(db.TempServiceResponse, {
        foreignKey: 'mockId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.ServiceResponse.belongsTo(db.Mock, {
        foreignKey: 'mockId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.TempServiceResponse.belongsTo(db.TempMock, {
        foreignKey: 'mockId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })

    db.User.hasMany(db.Token, {
        foreignKey: 'userId',
        foreignKeyConstraint: true,
        onDelete: 'cascade'
    })
    db.Token.belongsTo(db.User, {
        foreignKey: 'userId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.Project.hasMany(db.Token, {
        foreignKey: 'projectId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.Token.belongsTo(db.Project, {
        foreignKey: 'projectId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.Mock.hasMany(db.Token, { 
        foreignKey: 'mockId', 
        foreignKeyConstraint: true, 
        onDelete: 'cascade' 
    })
    db.Token.belongsTo(db.Mock, {
        foreignKey: 'mockId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })

    db.User.hasMany(db.ApiResponseSchemaType, {
        foreignKey: 'userId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.ApiResponseSchemaType.belongsTo(db.User, {
        foreignKey: 'userId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.Project.hasMany(db.ApiResponseSchemaType, {
        foreignKey: 'projectId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.ApiResponseSchemaType.belongsTo(db.Project, {
        foreignKey: 'projectId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.Mock.hasMany(db.ApiResponseSchemaType, {
        foreignKey: 'mockId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.ApiResponseSchemaType.belongsTo(db.Mock, {
        foreignKey: 'mockId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })

    db.User.hasMany(db.TempApiResponseSchemaType, {
        foreignKey: 'userId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.TempApiResponseSchemaType.belongsTo(db.User, {
        foreignKey: 'userId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.Project.hasMany(db.TempApiResponseSchemaType, {
        foreignKey: 'projectId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.TempApiResponseSchemaType.belongsTo(db.Project, {
        foreignKey: 'projectId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.TempMock.hasMany(db.TempApiResponseSchemaType, {
        foreignKey: 'mockId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })
    db.TempApiResponseSchemaType.belongsTo(db.TempMock, {
        foreignKey: 'mockId',
        foreignKeyConstraint: true,
        onDelete: 'cascade',
    })

    return db
}

module.exports = dbConn
