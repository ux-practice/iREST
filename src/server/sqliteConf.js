import logger from "./logger"

/* eslint-disable global-require */
const {Sequelize, DataTypes} = require('sequelize')

/* eslint-enable global-require */
const sequelize = new Sequelize(
  process.env.IREST_SQLITE_DATABASE,
  process.env.IREST_SQLITE_USERNAME,
  process.env.IREST_SQLITE_PASSWORD,
  {
    host: process.env.IREST_SQLITE_HOST,
    dialect:process.env.IREST_SQLITE_DIALECT /* one of 'mysql' | 'mariadb' | 'postgres' | 'mssql' */,
    storage: process.env.IREST_SQLITE_STORAGE,
    logging: false
  }
)

sequelize
  .authenticate()
  .then(() => {})
  .catch(error => {})

const db = {}
db.sequelize = sequelize
/* eslint-enable global-require */
require("./models/index")(db,sequelize, DataTypes)

db.sequelize.query("PRAGMA foreign_keys = false;")
db.sequelize.sync({force: false,alter:{drop: false}})
    .then((resp) => {

      db.sequelize.query("PRAGMA foreign_keys = true;")
    })
    .catch(error => {logger.error(error)})

export default db