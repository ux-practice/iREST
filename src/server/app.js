/* eslint-disable global-require */
import Express from 'express'
import webpack from 'webpack'
import webpackDevMiddleware from 'webpack-dev-middleware'
import webpackHotMiddleware from 'webpack-hot-middleware'
import winston, {format,transports} from 'winston'
import expressWinston from 'express-winston'
import cors from 'cors'
import path from 'path'
import saslprep from 'saslprep'
import compression from "compression"
import logger from './logger'

if (!saslprep) {
  logger.warn('Warning: no saslprep library specified. Passwords will not be sanitized')
}

// dot env config
const cwd = process.cwd()
const {NODE_ENV} = process.env

let dotEnvConfig
if (NODE_ENV === 'production') {
  dotEnvConfig = require('dotenv').config({path: `${cwd}/.env.prod`})
} else if (NODE_ENV === 'remote') {
  dotEnvConfig = require('dotenv').config({path: `${cwd}/.env.remote.prod`})
} else {
  dotEnvConfig = require('dotenv').config({path: `${cwd}/.env`})
}

const dotEnvExpand = require('dotenv-expand')

dotEnvExpand(dotEnvConfig)

const app = Express()

const corsOptions = {
  origin: '*',
}
app.use(cors(corsOptions))
app.use(compression())

const {printf,combine,timestamp,label,colorize} = format
// express winston acesss logs.before handling the routes.
if (!process.env.NODE_ENV === "production") {
  app.use(
    expressWinston.logger({
      transports: [
        new transports.Console(),
      ],
      format: combine(colorize(), combine(
        colorize({
          all: true,
        }),
        label({
          label: '\n[Access Log]',
        }),
        timestamp({
          format: 'YY-MM-DD HH:MM:SS',
        }),
        printf(
          // eslint-disable-next-line no-shadow
          ({label, timestamp, level, message, ...rest}) => `${label} ${timestamp} ${level}: ${message} ${JSON.stringify(rest)}`
        )
      )),
    })
  )
}

app.use(Express.json())
app.use(Express.urlencoded({extended: true}))
// const rawBodySaver = (req, res, buf, encoding) => {
//   if (buf && buf.length) {
//     req.rawBody = buf.toString(encoding || 'utf8')
//   }
// }

// app.use(bodyParser.json({verify: rawBodySaver})) // for parsing application/json
// app.use(bodyParser.urlencoded({verify: rawBodySaver, extended: true})) // for parsing application/x-www-form-urlencoded
// app.use(bodyParser.raw({verify: rawBodySaver, type: '*/*'})) // for parsing raw text data

// sqlite conf added
if (NODE_ENV !== 'test') {
  // eslint-disable-next-line global-require
  require("./sqliteConf")
}

// api route config
const {EXPRESS_REST_BASE_URL} = process.env
app.use(EXPRESS_REST_BASE_URL, require('./routes/api'))

if (NODE_ENV === 'development') {
  // eslint-disable-next-line global-require
  const config = require('../../config/webpack.dev')
  const compiler = webpack(config)

  app.use(
    webpackDevMiddleware(compiler, {
      publicPath: config.output.publicPath,
    })
  )
  app.use(webpackHotMiddleware(compiler))
  app.get('*', (req, res, next) => {
    const filename = path.join(compiler.outputPath, 'index.html')

    compiler.outputFileSystem.readFile(filename, (err, result) => {
      if (err) {
        return next(err)
      }
      res.set('content-type', 'text/html')
      res.send(result)
      res.end()
    })
    res.sendFile('index.html', {root: './public'})
  })
} else {
  // production config without webpack
  app.use(Express.static(process.env.MODULE_DIST_PATH))

  app.get('*', (req, res) => {
    res.sendFile('index.html', {root: process.env.MODULE_DIST_PATH})    
  })
}

// express winston error log
app.use(
  expressWinston.errorLogger({
    transports: [
      new winston.transports.Console({
        json: true,
        colorize: true,
      }),
    ],
  })
)


export default app
