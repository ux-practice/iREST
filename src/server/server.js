import {Server} from 'http'
import ChildProcess from 'child_process'
import app from './app'
import logger from './logger'


const server = new Server(app)
const port = process.env.PORT || 9000
const {NODE_ENV} = process.env

server.on('error', async (error) => {
  if (error.code === 'EADDRINUSE') {
    logger.error(`Port ${port} is already in use!`)
  }
  
})

server.listen(port, err => {
  if (err) {
    return logger.debug(err)
  }
  if (NODE_ENV === "production") {
    const start = (process.platform === 'darwin'? 'open': process.platform === 'win32'? 'start': 'xdg-open')
    ChildProcess.exec(`${start} http://localhost:${port}`)
  }
  return logger.http(`iRest Server is listening on localhost:${port}, open your browser on http://localhost:${port}`)
})
module.exports = server