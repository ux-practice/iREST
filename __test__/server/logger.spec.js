import {customFormat} from '../../src/server/logger'

describe('logger', () => {
  it('custom format no extra params', () => {
    const params = {
      label: 'Default Logger',
      timestamp: '2021-02-03T10:50:25.713+05:30',
      level: 'debug',
      message: 'hello world!',
    }
    expect(customFormat(params)).toMatchSnapshot()
  })
  it('custom format no extra params', () => {
    const params = {
      label: 'Default Logger',
      timestamp: '2021-02-03T10:50:25.713+05:30',
      level: 'debug',
      message: 'hello world!',
      headers: {'content-type': 'application/json'}
    }
    expect(customFormat(params)).toMatchSnapshot()
  })
})
