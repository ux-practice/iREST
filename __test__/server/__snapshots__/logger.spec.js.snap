// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger custom format no extra params 1`] = `"Default Logger 2021-02-03T10:50:25.713+05:30 debug: hello world! "`;

exports[`logger custom format no extra params 2`] = `"Default Logger 2021-02-03T10:50:25.713+05:30 debug: hello world! {\\"headers\\":{\\"content-type\\":\\"application/json\\"}}"`;
